File: parsers/src/test/java/wcl/gpromanager/parsers/tracks/TrackParserTest.java
package wcl.gpromanager.parsers.tracks;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import wcl.gpromanager.parsers.testutils.HtmlFileLoader;
import wcl.gpromanager.parsers.tracks.entities.Track;
import wcl.gpromanager.parsers.utils.FloatValueRetriever;
import wcl.gpromanager.parsers.utils.IntegerValueRetriever;
import wcl.gpromanager.parsers.utils.StringValueRetriever;

import java.util.ArrayList;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
// @TODO COnvert to use AssertJ?
public class TrackParserTest {
    private static final String TRACK_TEST_HTML_FILE
            = "/htmlpages/Tracks/Track/Track information - Buenos Aires - Grand Prix Racing Online 26-0-24.html";

    private TrackParser testParser;
    private Track testTrack;

    @BeforeEach
    public void setupTestParser() {
        testParser = new TrackParser(new FloatValueRetriever(), new IntegerValueRetriever(), new StringValueRetriever());
        testTrack = testParser.parse(2, HtmlFileLoader.loadHtmlFile(TRACK_TEST_HTML_FILE));
    }

    @Test
    public void whenTrackListPageParsed_thenIdCorrectlyExtracted() {
        assertEquals(2, testTrack.id());
    }

    @Test
    public void whenTrackListPageParsed_thenTrackNameCorrectlyExtracted() {
        assertEquals("Buenos Aires", testTrack.trackName());
    }

    @Test
    public void whenTrackListPageParsed_thenLocationCorrectlyExtracted() {
        assertEquals("Argentina", testTrack.location());
    }


    @Test
    public void whenTrackListPageParsed_thenImageFileNameCorrectlyExtracted() {
        assertEquals("track_buenos_arg.gif", testTrack.imageFileName());
    }
    @Test
    public void whenTrackListPageParsed_thenRaceDistanceCorrectlyExtracted() {
        assertEquals(306.6, testTrack.raceDistance(), 0.1);
    }

    @Test
    public void whenTrackListPageParsed_thenLapsCorrectlyExtracted() {
        assertEquals(72, testTrack.laps());
    }

    @Test
    public void whenTrackListPageParsed_thenLapDistanceCorrectlyExtracted() {
        assertEquals(4.258, testTrack.lapDistance(), 0.001);
    }

    @Test
    public void whenTrackListPageParsed_thenAverageSpeedCorrectlyExtracted() {
        assertEquals(189.63, testTrack.averageSpeed(), 0.01);
    }

    @Test
    public void whenTrackListPageParsed_thenCornersCorrectlyExtracted() {
        assertEquals(16, testTrack.corners());
    }

    @Test
    public void whenTrackListPageParsed_thenPitLaneTimeCorrectlyExtracted() {
        assertEquals(19.5, testTrack.pitlaneTime(), 0.01);
    }

    @Test
    public void whenTrackListPageParsed_thenPowerCorrectlyExtracted() {
        assertEquals(12, testTrack.power());
    }

    @Test
    public void whenTrackListPageParsed_thenHandlingCorrectlyExtracted() {
        assertEquals(11, testTrack.handling());
    }

    @Test
    public void whenTrackListPageParsed_thenAccelerationCorrectlyExtracted() {
        assertEquals(9, testTrack.acceleration());
    }

    @Test
    public void whenTrackListPageParsed_thenDownforceCorrectlyExtracted() {
        assertEquals("High", testTrack.downforce());
    }

    @Test
    public void whenTrackListPageParsed_thenOvertakingCorrectlyExtracted() {
        assertEquals("Very hard", testTrack.overtaking());
    }

    @Test
    public void whenTrackListPageParsed_thenSuspensionRigidityCorrectlyExtracted() {
        assertEquals("Hard", testTrack.suspensionRigidity());
    }

    @Test
    public void whenTrackListPageParsed_thenFuelConsumptionCorrectlyExtracted() {
        assertEquals("Very low", testTrack.fuelConsumption());
    }

    @Test
    public void whenTrackListPageParsed_thenTyreWearCorrectlyExtracted() {
        assertEquals("High", testTrack.tyreWear());
    }

    @Test
    public void whenTrackListPageParsed_thenGripLevelCorrectlyExtracted() {
        assertEquals("Normal", testTrack.gripLevel());
    }
}

File: parsers/src/test/kotlin/wcl/gpromanager/parsers/tracks/TrackParserTest.kt
package wcl.gpromanager.parsers.tracks

import org.junit.jupiter.api.*
import wcl.gpromanager.parsers.testutils.HtmlFileLoader.loadHtmlFile
import wcl.gpromanager.parsers.utils.*
import org.assertj.core.api.Assertions.*
import wcl.gpromanager.entities.track.Track

class TrackParserTest {
    private lateinit var testParser: TrackParser
    private lateinit var testTrack: Track

    @BeforeEach
    fun setupTestParser() {
        testParser = TrackParser(FloatValueRetriever(), IntegerValueRetriever(), StringValueRetriever())
        testTrack = testParser.parse(2, loadHtmlFile(TRACK_TEST_HTML_FILE))
    }

    @Test
    fun whenTrackListPageParsed_thenIdCorrectlyExtracted() {
        assertThat(testTrack.id).isEqualTo(2)
    }

    @Test
    fun whenTrackListPageParsed_thenTrackNameCorrectlyExtracted() {
        assertThat(testTrack.trackName).isEqualTo("Buenos Aires")
    }

    @Test
    fun whenTrackListPageParsed_thenLocationCorrectlyExtracted() {
        assertThat(testTrack.location).isEqualTo("Argentina")
    }

    @Test
    fun whenTrackListPageParsed_thenImageFileNameCorrectlyExtracted() {
        assertThat(testTrack.imageFileName).isEqualTo("track_buenos_arg.gif")
    }

    @Test
    fun whenTrackListPageParsed_thenRaceDistanceCorrectlyExtracted() {
        assertThat(testTrack.raceDistance).isEqualTo(306.6f, withPrecision(0.1f))
    }

    @Test
    fun whenTrackListPageParsed_thenLapsCorrectlyExtracted() {
        assertThat(testTrack.laps).isEqualTo(72)
    }

    @Test
    fun whenTrackListPageParsed_thenLapDistanceCorrectlyExtracted() {
        assertThat(testTrack.lapDistance).isEqualTo(4.258f, withPrecision(0.001f))
    }

    @Test
    fun whenTrackListPageParsed_thenAverageSpeedCorrectlyExtracted() {
        assertThat(testTrack.averageSpeed).isEqualTo(189.63f, withPrecision(0.01f))
    }

    @Test
    fun whenTrackListPageParsed_thenCornersCorrectlyExtracted() {
        assertThat(testTrack.corners).isEqualTo(16)
    }

    @Test
    fun whenTrackListPageParsed_thenPitLaneTimeCorrectlyExtracted() {
        assertThat(testTrack.pitlaneTime).isEqualTo(19.5f, withPrecision(0.1f))
    }

    @Test
    fun whenTrackListPageParsed_thenPowerCorrectlyExtracted() {
        assertThat(testTrack.power).isEqualTo(12)
    }

    @Test
    fun whenTrackListPageParsed_thenHandlingCorrectlyExtracted() {
        assertThat(testTrack.handling).isEqualTo(11)
    }

    @Test
    fun whenTrackListPageParsed_thenAccelerationCorrectlyExtracted() {
        assertThat(testTrack.acceleration).isEqualTo(9)
    }

    @Test
    fun whenTrackListPageParsed_thenDownforceCorrectlyExtracted() {
        assertThat(testTrack.downforce).isEqualTo("High")
    }

    @Test
    fun whenTrackListPageParsed_thenOvertakingCorrectlyExtracted() {
        assertThat(testTrack.overtaking).isEqualTo("Very hard")
    }

    @Test
    fun whenTrackListPageParsed_thenSuspensionRigidityCorrectlyExtracted() {
        assertThat(testTrack.suspensionRigidity).isEqualTo("Hard")
    }

    @Test
    fun whenTrackListPageParsed_thenFuelConsumptionCorrectlyExtracted() {
        assertThat(testTrack.fuelConsumption).isEqualTo("Very low")
    }

    @Test
    fun whenTrackListPageParsed_thenTyreWearCorrectlyExtracted() {
        assertThat(testTrack.tyreWear).isEqualTo("High")
    }

    @Test
    fun whenTrackListPageParsed_thenGripLevelCorrectlyExtracted() {
        assertThat(testTrack.gripLevel).isEqualTo("Normal")
    }

    companion object {
        private const val TRACK_TEST_HTML_FILE = "/htmlpages/Tracks/Track/Track information - Buenos Aires - Grand Prix Racing Online 26-0-24.html"
    }
}